node/uhpo/src/crypto/signature.rs

```diff
@@ -0,0 +1,269 @@
+        self.add_xonly_tweak(secp, tweak)
+    }
+
+    fn to_keypair(&self) -> Keypair {
+        self.clone()
+    }
+}
+
+impl Secp256k1Behavior for Secp256k1<All> {
+    fn sign_schnorr_with_rng<R: Rng + CryptoRng>(
+        &self,
+        msg: &Message,
+        keypair: &Keypair,
+        rng: &mut R,
+    ) -> Signature {
+        self.sign_schnorr_with_rng(msg, keypair, rng)
+    }
+
+    fn verify_schnorr(
+        &self,
+        signature: &Signature,
+        message: &Message,
+        pubkey: &XOnlyPublicKey,
+    ) -> Result<(), bitcoin::secp256k1::Error> {
+        self.verify_schnorr(signature, message, pubkey)
+    }
+}
+
+/// Adds a signature to a transaction input.
+///
+/// This function creates a signature for the specified input of a transaction,
+/// optionally applying a tweak to the private key before signing.
+pub fn add_signature<S, K, E>(
+    transaction: &mut Transaction,
+    input_idx: usize,
+    prevouts: &[&TxOut],
+    private_key: S,
+    tweak: Option<&Scalar>,
+    secp: &E,
+) -> Result<(), UhpoError>
+where
+    S: SecretKeyBehavior<K, E>,
+    K: KeypairBehavior<E>,
+    E: Secp256k1Behavior + 'static,
+{
+    // apply tweak if provided
+    let keypair: Keypair = match tweak {
+        Some(tweak) => private_key
+            .keypair(secp)
+            .add_xonly_tweak(secp, tweak)
+            .map_err(UhpoError::KeypairCreationError)?,
+        None => private_key.keypair(secp).to_keypair(),
+    };
+
+    // Compute the sighash
+    let mut sighash_cache = SighashCache::new(transaction.clone());
+    let sighash = sighash_cache.taproot_key_spend_signature_hash(
+        input_idx,
+        &Prevouts::All(prevouts),
+        TapSighashType::All,
+    )?;
+
+    // Create and sign the message
+    let message = Message::from_digest(sighash.as_raw_hash().to_byte_array());
+    let signature = secp.sign_schnorr_with_rng(&message, &keypair, &mut rand::thread_rng());
+
+    // Verify the signature
+    secp.verify_schnorr(&signature, &message, &keypair.x_only_public_key().0)
+        .map_err(UhpoError::SignatureVerificationError)?;
+
+    // Add the signature to the transaction input
+    let mut vec_sig = signature.serialize().to_vec();
+    vec_sig.push(0x01);
+    transaction.input[input_idx].witness.push(vec_sig);
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod mock_tests {
+    use crate::crypto::signature::*;
+
+    use bitcoin::{absolute::LockTime, transaction::Version, Amount, ScriptBuf, TxIn};
+    use rand::rngs::ThreadRng;
+
+    // dummy transaction for testing with no inputs
+    pub fn create_dummy_transaction() -> Transaction {
+        Transaction {
+            version: Version::TWO,
+            lock_time: LockTime::ZERO,
+            input: vec![],
+            output: vec![TxOut {
+                value: Amount::from_sat(50000000),
+                script_pubkey: ScriptBuf::new(),
+            }],
+        }
+    }
+
+    #[test]
+    fn test_add_signatures_key_creation_fails() {
```

Comment:
Let's rename test to 

`test_add_signature_with_invalid_tweak_error_signature_creation_should_fail`

basically

`test_<fn>_<under_which_conditions>_<what_should_happen>`